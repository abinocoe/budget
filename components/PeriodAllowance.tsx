import React, { Component } from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { Input } from "react-native-elements";

import * as colours from "../constants/colours";

interface Props {
  allowanceAmount: number;
  updateAllowance: (amount: number) => void;
}

interface State {
  showInput: boolean;
  value: string;
  rawValue: string;
}

class PeriodAllowanceInput extends Component<Props> {
  public state: Readonly<State> = {
    showInput: false,
    value: "",
    rawValue: ""
  };

  public amountChanged(inputText: string) {
    let tmpAmount = "";
    const tmpValue = inputText.slice(-1);
    const newRawValue = this.state.rawValue + tmpValue;

    if (this.state.value.length > inputText.length) {
      this.setState({
        value: "",
        rawValue: ""
      });
    } else {
      if (newRawValue.length === 1) {
        tmpAmount = `0.0${newRawValue}`;
      } else if (newRawValue.length === 2) {
        tmpAmount = `0.${newRawValue}`;
      } else {
        tmpAmount = inputText
          .replace(".", "")
          .replace(/(\d{2})\b$/, ".$1")
          .replace(/^0(\d.\d{2})/, "$1");
      }

      this.setState({
        value: tmpAmount,
        rawValue: newRawValue
      });
    }
  }

  public render() {
    const { updateAllowance, allowanceAmount } = this.props;
    return (
      <View>
        {!this.state.showInput ? (
          <TouchableOpacity
            onPress={() => this.setState({ showInput: !this.state.showInput })}
          >
            <View style={{ alignItems: "center", justifyContent: "center" }}>
              <Text
                style={{
                  fontSize: 30,
                  paddingHorizontal: 5,
                  marginTop: 10,
                  marginBottom: 5,
                  borderRadius: 5,
                  backgroundColor: colours.pink,
                  width: "90%",
                  textAlign: "center",
                  color: "white",
                  elevation: 3
                }}
              >
                {isNaN(allowanceAmount)
                  ? ""
                  : (allowanceAmount / 100).toFixed(2)}
              </Text>
            </View>
          </TouchableOpacity>
        ) : (
          <View style={{ alignItems: "center" }}>
            <Input
              testID="input"
              autoFocus={true}
              containerStyle={{
                marginTop: 10,
                marginBottom: 5,
                backgroundColor: "pink",
                width: "90%",
                borderRadius: 5,
                elevation: 2,
                height: 40
              }}
              style={{ color: "white" }}
              inputStyle={{
                color: "white",
                textAlign: "right",
                fontSize: 30,
                padding: 0,
                paddingRight: 5,
                borderColor: "transparent"
              }}
              placeholderTextColor="white"
              inputContainerStyle={styles.inputContainerStyle}
              keyboardType="number-pad"
              onChangeText={text => {
                this.amountChanged(text);
              }}
              onEndEditing={e => {
                this.setState({ showInput: !this.state.showInput });
                if (e.nativeEvent.text !== "") {
                  const intAmount = e.nativeEvent.text.replace(".", "");
                  updateAllowance(parseInt(intAmount, 10));
                  this.setState({
                    value: "",
                    rawValue: ""
                  });
                }
              }}
              placeholder={
                isNaN(allowanceAmount)
                  ? "0.00"
                  : (allowanceAmount / 100).toFixed(2)
              }
              value={this.state.value}
            />
          </View>
        )}
        <Text style={styles.text}>Total amount this month</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  inputContainerStyle: {
    borderBottomColor: "rgba(0,0,0,0)"
  },
  text: {
    fontSize: 12
  }
});

export default PeriodAllowanceInput;
