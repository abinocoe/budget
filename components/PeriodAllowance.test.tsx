import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import PeriodAllowance from "./PeriodAllowance";

describe("PeriodAllowance", () => {
  it("renders correctly", () => {
    const output = shallow(
      <PeriodAllowance allowanceAmount={50000} updateAllowance={() => null} />
    );
    expect(toJson(output)).toMatchSnapshot();
  });

  it("renders input correctly", () => {
    const output = shallow(
      <PeriodAllowance allowanceAmount={50000} updateAllowance={() => null} />
    );
    output.setState({ showInput: true });
    expect(toJson(output)).toMatchSnapshot();
  });

  it("calls updateAllowance on input submit", () => {
    const upAllSpy = jest.fn();
    const output = shallow(
      <PeriodAllowance allowanceAmount={50000} updateAllowance={upAllSpy} />
    );
    output.setState({ showInput: true });
    const input = output.findWhere(node => {
      return node.prop("testID") === "input";
    });
    input.prop("onEndEditing")({ nativeEvent: { text: "111" } });
    expect(upAllSpy).toHaveBeenCalledWith(111);
  });
});
