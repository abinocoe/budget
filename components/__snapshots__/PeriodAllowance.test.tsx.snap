// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeriodAllowance renders correctly 1`] = `
<View>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#EC407A",
            "borderRadius": 5,
            "color": "white",
            "elevation": 3,
            "fontSize": 30,
            "marginBottom": 5,
            "marginTop": 10,
            "paddingHorizontal": 5,
            "textAlign": "center",
            "width": "90%",
          }
        }
      >
        500.00
      </Text>
    </View>
  </TouchableOpacity>
  <Text
    style={
      Object {
        "fontSize": 12,
      }
    }
  >
    Total amount this month
  </Text>
</View>
`;

exports[`PeriodAllowance renders input correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <ForwardRef(Themed.Input)
      autoFocus={true}
      containerStyle={
        Object {
          "backgroundColor": "pink",
          "borderRadius": 5,
          "elevation": 3,
          "height": 40,
          "marginBottom": 5,
          "marginTop": 10,
          "width": "90%",
        }
      }
      inputContainerStyle={
        Object {
          "borderBottomColor": "rgba(0,0,0,0)",
        }
      }
      inputStyle={
        Object {
          "borderColor": "transparent",
          "color": "white",
          "fontSize": 30,
          "padding": 0,
          "paddingRight": 5,
          "textAlign": "right",
        }
      }
      keyboardType="number-pad"
      onChangeText={[Function]}
      onEndEditing={[Function]}
      placeholder="500.00"
      placeholderTextColor="white"
      style={
        Object {
          "color": "white",
        }
      }
      testID="input"
      value=""
    />
  </View>
  <Text
    style={
      Object {
        "fontSize": 12,
      }
    }
  >
    Total amount this month
  </Text>
</View>
`;
